### ISSUE TITLE
[DOCS] Use Cookie to save Token while user is logged in

### ISSUE THREAD

**Issue by mrwhy-orig** [tymondesigns/jwt-auth#1594]
### Subject of the issue
Describe your issue here.

### Your environment
| Q                 | A
| ----------------- | ---
| Bug?              | no
| New Feature?      | no / yes
| Framework         | Laravel
| Framework version | 5.6
| Package version   | 1.0.0-rc.2
| PHP version       | 7.1.3

A short how to use jwt with cookies to achieve reauthentication would be nice.

**Comment by krott** [comment]
## Update:
Ok, I think I found the problem. At least it's working for me now. The root of the culprit I guess is that Lumen by design no longer does cookies which I find a bit of a flaw in the light of all the blogs and OWASP suggestions of not storing a JWT in localstorage but rather in a httponly cookie to prevent XSS and deal with CSRF accordingly.
So, the jwt-auth doesn't include the cookie parser with the LumenServiceProvider which is what you register in app.php as a service provider:
```
$app->register(Tymon\JWTAuth\Providers\LumenServiceProvider::class);
```

So when you add 
```
use Tymon\JWTAuth\Http\Parser\Cookies;
```
to the top of _jwt-auth\src\Providers\LumenServiceProvider.php_ and add 
```
new Cookies($this->config('decrypt_cookies'))
```
into the array at the very end of the file

```
$this->app['tymon.jwt.parser']->setChain([<br>
            new AuthHeaders,
            new QueryString,
            new InputSource,
            new LumenRouteParams,
            new Cookies($this->config('decrypt_cookies')),
        ]);
```
then you should be able use the cookie authentication in Lumen as well.

### Original post:
+1

Everything seems to be working fine when I use the Authorization header with the Bearer token, but the moment I switch to cookies, I get an 'Unauthorized.' while trying to make any subsequent ajax call after the log in. I can see the cookie being set in the browser upon login and the cookie being included in the request header of subsequent XHR calls but I guess there needs to be some config changes to make jwt-auth read the cookie? My setup is as described in docs on http://jwt-auth.com running Lumen 5.6.

**Comment by mrwhy-orig** [comment]
Unfortunately I'm using Laravel. For Laravel the AbstractServiceProvider Class registers the TokenParsers and therin the Cookies Parser is defined. As I can see the KeyTrait tells the Cookies parser the name of the cookie to look for: "token".
In the jwt config file, i guess, i have to use decrypt_cookies => true
But i think i miss something, if i make a page reload while logged in I get an unathenticated 401 error, but the cookie is still in place.

**Comment by krott** [comment]
Yes, for Laravel it sounds like it's working out of the box.
So let's start at the beginning because you have not provided much in regards to the environment you're working with other than Laravel:

- so you are able to log in, i.e. you get a token and see a page containing user information of the logged in user?
- reload the page you hit the the Auth middleware in Laravel and you get the dreaded 401: can you confirm that the cookie is actually sent as part of the request header? Maybe the domain is not set correctly on the cookie...
- you're probably right with the decrypt_cookie part or you could try and add the cookie name to the exception list in the EncryptCookies middleware in Laravel. I don't really see much benefit of encrypting the JWT unless you put all kinds of sensitive information in there. You should still protect against CSRF when using a JWT cookie.

**Comment by mrwhy-orig** [comment]
1. Yes, 2. The Domain is correctly set, 3. Right - Yesterday I discovered another error. If I set decrypt_cookie to true, and name the cookie like the default KeyTrait name, i got an "payload is invalid" error on login after the page reload. So i guess something is wrong within the cookie itself.
I'm sending the cookie along with token array in json format. Is it possible, that the token is restricted to one auth method ?
Or do I have to set the remember_token in the user table manually? Sorry for the noob questions, i just started with Laravel. Thanks for your help!

**Comment by krott** [comment]
Ok, to keep it simple: try it without encrypting the cookie and see if that works. I personally never used that functionality in Laravel but I heard that sometimes there are special characters that can encrypt into something like %3D and that can cause a problem during decryption....but maybe that's fixed in v5.6, no idea, can't help you there. Personally, I don't put anything sensitive in my JWTs so encrypting the cookie is a mood point, i.e. if you're going to steal/hijack the cookie and send it to my server, it will still get decrytped server side and it works. It's a token...anyways, back to the topic on hand.

Not sure what you mean by 

> I'm sending the cookie along with token array in json format

Sounds like you're sending the cookie and a payload containing the token? Why? Any particular reason for sending it twice? 
If I'm reading this right, in jwt-auth\src\Http\Parser\Parser.php line 88 the parseToken() method goes through the list of parsers until it finds a good response then returns immediately - but the cookie is last in the chain, so....but just conceptually, I don't see any benefit other than testing of having the token sent back via multiple methods. jwt-auth doesn't care from the looks of it instead it takes the first match and runs with it. But if I were to send it back in anything other than a cookie I'd go the Authorization header route with a Bearer token just because my payload then can be whatever I choose it to be without having to do any extra work on the server side, taking it apart to get the bits and pieces I want for some request.

Remember_token? I know Laravel is providing that functionality with their Session Auth Guard but you're using JWTs, so that doesn't work. When you make the call to auth->attempt(<credentials>) during your login, you're calling the attempt() method of jwt-auth which doesn't do remember me tokens. You set a TTL for the token and JWT_REFRESH_TTL (check the jwt config file) and you'll have to code something to provide that "remember me" functionality from there I guess. But either you're using Laravel sessions for authentication and session management or JWT, I don't think it makes sense to have both but I'm happy to learn.

**Comment by mrwhy-orig** [comment]
Ok, then... now I figured it out. Thanks @krott.

I'm using @websanova/vue-auth for authentication with vue SPA in Laravel. The main thing was, that I had to alter the response headers in Laravel. If you add the Authorization Part to the header like 
`$responseObj->withHeaders(['Authorization' => "Bearer " . $token])` The header is set properly. VueAuth (standard Bearer Auth driver) is looking for this header part to get and store the token.

I'm wondering why all the tutorials out there (those i read) are recomending to send the token in a plain json response. They never mentioned to alter the header or altering the auth driver... 

Anyway, if somebody needs more info on that just ask...
